### ISSUE TITLE
The LiveMigration fails if some RAM stress test activated during the LiveMigration time.

### ISSUE THREAD

**Issue by harutyundermenjyan** [kubevirt/kubevirt#3504]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind enhancement

**What happened**:
The LiveMigration never succeeded if some RAM stress test activated during the LiveMigration time, in particular:
`stress -m 1 --vm-bytes 30G` _or_ `stress -m 1 --vm-bytes 30G --vm-keep`
Also, noticed if the --vm-bytes larger than 200M then the migration can't be completed.
The tests are performed on with 256G ram machines.

**What you expected to happen**:
In a normal case, it should be migrated to a target node and the old pod on the source node should be completed.
**How to reproduce it (as minimally and precisely as possible)**:
Create a VM with container disk and perform live migration with the aforementioned ram stress test. 

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`): `v0.28.0`
- Kubernetes version (use `kubectl version`): `client v1.17.0`, `server v1.17.4`
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`): `4.15.0-99-generic #100-Ubuntu SMP Wed Apr 22 20:32:56 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
- Others:
Also sharing the source and target virt-launcher and handler logs.
[source-virt-handler.log](https://github.com/kubevirt/kubevirt/files/4719960/source-virt-handler.log)
[source-virt-launcher.log](https://github.com/kubevirt/kubevirt/files/4719961/source-virt-launcher.log)
[target-virt-handler.log](https://github.com/kubevirt/kubevirt/files/4719962/target-virt-handler.log)
[target-virt-launcher.log](https://github.com/kubevirt/kubevirt/files/4719963/target-virt-launcher.log)

**Comment by rmohr** [comment]
In general I think if your migraitons can succeed in such case probably depend on the networkbandwidth. If you dirty the ram faster than you can trasfer the state over the network, I don't think that the migration can succeed by default.

@vladikr you implemented an autoconverge feature, right? Maybe that could help here.

**Comment by harutyundermenjyan** [comment]
Actually, the machines have 2 interfaces each of them 10gbit/s, also they bonded and the full capability is 20gbit/s, I think the bandwidth is good enough in my case and it shouldn't be the cause.

**Comment by vladikr** [comment]
I doubt that the virt-launcher has access to such a large bandwidth of 20gbit/s. Also, please keep in mind that we currently proxy the migration process from the virt-launcher to virt-handler and back on the destination side. This creates a large overhead.

At this moment live migration shouldn't be viewed as a high-performance feature. In the future, we are planning to enhance the process. Perhaps providing a dedicated network or/and post live-migration.
As Roman mentioned, It's possible to try to allow live migration to converge automatically. You can update the kubevirt-config  `kubectl edit configmap kubevirt-config -n kubevirt`
and set `allowAutoConverge": true` under `migrations`
As in: https://kubevirt.io/user-guide/#/installation/live-migration?id=changing-cluster-wide-migration-limits

**Comment by harutyundermenjyan** [comment]
After adding flag `allowAutoConverge: "true"` got this error:
`Error migrating VirtualMachine Internal error occurred: admission webhook "migration-create-validator.kubevirt.io" denied the request: LiveMigration feature gate is not enabled in kubevirt-config
`
But actually the LiveMigration future-gate is enabled.
Here is the config:
```k get cm kubevirt-config -n kubevirt -o yaml
apiVersion: v1
data:
  debug.useEmulation: "false"
  feature-gates: DataVolumes,LiveMigration
  migrations: |-
    parallelMigrationsPerCluster: 5
    parallelOutboundMigrationsPerNode: 2
    bandwidthPerMigration: 20Gi
    #completionTimeoutPerGiB: 1600
    #progressTimeout: 800
    allowAutoConverge: "true"
kind: ConfigMap
metadata:
  creationTimestamp: "2020-05-20T11:15:55Z"
  name: kubevirt-config
  namespace: kubevirt
  resourceVersion: "32224697"
  selfLink: /api/v1/namespaces/kubevirt/configmaps/kubevirt-config
  uid: 83abbca9-1cd5-4647-b92a-c0941ca50b4d
```
When I delete allowAutoConverge: "true" flag it the migration starts

**Comment by harutyundermenjyan** [comment]
The flag `allowAutoConverge: true` didn't help.
I tried LiveMigrate with this: `stress -m 1 --vm-bytes 1024M --vm-keep`.
But an interesting thing is that I tried before to do LiveMigration with Redis with about ~100G ram usage and the migration succeeded about after 25 minutes, but with the aforementioned stress test it didn't work.

**Comment by harutyundermenjyan** [comment]
After adding the flag `allowAutoConverge: true` managed to migrate 4096M `--vm-bytes 4096M` (instead of 200M) but no more.
Interesting thing is that when I'm migrating Redis-server with almost 100G ram usage it works.
Just wondering what could be the issue, maybe it relates to the intensity of read-write?

**Comment by rmohr** [comment]
Oh, now I think I know what the issue is. We have another config setting `bandwidthPerMigration` which is set to `64Mi` by default. This means that every migration has a bandwidth limitation. It is deliberately pretty low to not by accident make production workloads unstable. Can you try making this much higher?

**Comment by harutyundermenjyan** [comment]
Unfortunately seems the `bandwidthPerMigration` not the cause in my case, all my tests I did with `bandwidthPerMigration:` `20Gi`
Below is my kubevirt-config:
```
k get cm kubevirt-config -n kubevirt -o yaml --export
Flag --export has been deprecated, This flag is deprecated and will be removed in future.
apiVersion: v1
data:
  debug.useEmulation: "false"
  feature-gates: DataVolumes,LiveMigration
  migrations: |-
    parallelMigrationsPerCluster: 5
    parallelOutboundMigrationsPerNode: 2
    bandwidthPerMigration: 20Gi
    #completionTimeoutPerGiB: 1600
    #progressTimeout: 800
    #allowAutoConverge: true
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"debug.useEmulation":"false","feature-gates":"DataVolumes,LiveMigration","migrations":"parallelMigrationsPerCluster: 5\nparallelOutboundMigrationsPerNode: 2\nbandwidthPerMigration: 20Gi\n#completionTimeoutPerGiB: 1600\n#progressTimeout: 800\nallowAutoConverge: true"},"kind":"ConfigMap","metadata":{"annotations":{},"name":"kubevirt-config","namespace":"kubevirt"}}
  creationTimestamp: null
  name: kubevirt-config
  selfLink: /api/v1/namespaces/kubevirt/configmaps/kubevirt-config
```
Could you please reproduce it on your side to see if you can manage the migration with ram stress test with 8G or more?
e.g. `stress -m 1 --vm-bytes 16G --vm-keep`

**Comment by rmohr** [comment]
@harutyundermenjyan I did not have too close a look on the conversation. @vladikr pointed me now to the fact that you are polluting 16G per second. Note that the autoconverge feature is throttling the CPU. If `stress` does the ram writes without much CPU involved, autoconverge does not really have much effect. I am not sure if 16G is now Gbyte or Gbit, but in both cases you are also already pretty close to your overall capacity and further, we unfortunately have  to do some proxying for the migrations, so I don't think that you can expect right now that VMs with such high memory write rates can migrate. Basically what @vladikr already said.

Let's start fresh: What is your target rate which you have to migrate?

**Comment by harutyundermenjyan** [comment]
In this case, 16G means Gbyte by basing on `stress` doc (`malloc B bytes per vm worker`).
The actual rate I didn't measure but anyway it depends on ram (`DIMM DDR4 Synchronous 2400 MHz (0.4 ns)`) and it should be faster and the network max capacity is 20gbit/s with bonding and 10gbit/s without bonding, so at least 10gbit/s is accessible in case if k8s overlay doesn't use the bond.
And also can I clarify when @vladikr mentioned that he doubts that virt-launcher has access to such a large bandwidth of 20gbit/s what are these doubts connected with?

**Comment by harutyundermenjyan** [comment]
One of my targets that have a few Redis servers with large ram usage and high read/write, so I want to test and understand how the kubevirt VM can manage this and can perform a live migration in case of need.
So as I understand it will work if all read and write perform via network because in that case the read/write can't be higher than the migration if they in the same network and have the same network capacity.
Do I notice right?

**Comment by harutyundermenjyan** [comment]
@rmohr @vladikr

**Comment by rmohr** [comment]
I think you understand the limitations right. I would still recommend to check it with the actual target workload. For instance `allowAutoConverge` is way more efficient if the workload uses a lot of CPU, since the auto converge feature basically throttles the CPU which will in such a case then slow down the memory writes. `stress` does not need much CPU to do memory writes, therefore `allowAutoConverge` may not have much effect.

Long story short, it may be that you can successfully migrate very busy redis VMs although `stress`- based tests may fail.

**Comment by harutyundermenjyan** [comment]
It's not a problem in terms of `allowAutoConverge` if Redis also uses a single CPU thread?

**Comment by rmohr** [comment]
@harutyundermenjyan I can't say much mure than trying out if it works good enough for you. `allowAutoConverge` works on real applications very often as expected but not on all applications. It is just something which helps a migration to have a higher chance to succeed.
